### ISSUE TITLE
PEP420 namespaces aren't recognized

### ISSUE THREAD

**Issue by charlesnicholson** [fpgmaas/deptry#740]
In our Python ecosystem, we use PEP420 namespace packages to prefix all of our packages with `companyname` to avoid package name collisions with third-party packages. For example, our directory structures might look like:

```
src/
  company.foo/
    company/
      foo/
        foo.py
    pyproject.toml
```

The pyproject.toml file is pretty vanilla:
```
[project]
name = "company.foo"
version = "0.0.1"
dependencies = [ ... ]

[build-system]
requires = ["setuptools", "wheel"]
build-backend = "setuptools.build_meta"
```

When running `deptry .` from the terminal with the current working directory set to `src/company.foo`, I get this error:

```
Assuming the corresponding module name of package 'company' is 'company'. Install the package or configure a package_module_name_map entry to override this behaviour.
```

This warning makes it seem like `deptry` can't understand these kinds of packages yet, since the package names aren't ambiguous and are, as best I can tell, conforming to https://peps.python.org/pep-0420/#nested-namespace-packages.

OS: macOS Sonoma 14.5
Language Version: Python 3.11
Poetry version: we don't use poetry. (We use python and uv, but uv plays no role in our package source layouts).

Thanks for `deptry`! Hopefully it will help us stop shipping broken wheel sets... :-|

**Comment by mkniewallner** [comment]
This should be fixed in [0.16.2](https://github.com/fpgmaas/deptry/releases/tag/0.16.2), if you want to check.

**Comment by charlesnicholson** [comment]
Thanks for looking into it! It still seems to be happening, unfortunately. I threw together a minimal reproduction in a public repo that you can clone and investigate, if you have the appetite for it:

https://github.com/charlesnicholson/deptry-pep420-repro

Also note the commented-out line in the `repro.sh`, is it possible that the `ROOT` argument isn't working as expected? Apologies if I'm misunderstanding the intended usage.

**Comment by mkniewallner** [comment]
Thanks for taking the time to create a reproduction repository, really appreciate it!

I won't be able to take a look right away, but will be during the weekend. In the meantime, I'll re-open the issue as it is not yet fixed.

**Comment by charlesnicholson** [comment]
Happy to! I set up a manual CI trigger for it, you can see it happening on `ubuntu-latest` here:

https://github.com/charlesnicholson/deptry-pep420-repro/actions/runs/9810619739/job/27091137978
<img width="1264" alt="Screenshot 2024-07-05 at 11 35 51 AM" src="https://github.com/fpgmaas/deptry/assets/3010295/22a0a059-073c-4dac-9f85-a2fd6086e1fd">

**Comment by mkniewallner** [comment]
(closed by mistake by linking a PR to this issue)

**Comment by kojiromike** [comment]
I'm sad that #753 was reverted. Deptry has never been anything close to our slowest performer in our CI suites so we wouldn't have minded if the performance had declined even quite a bit. Plus, all our codebases are small. Would it be possible to make it enabled by configuration, but disabled by default?

**Comment by mkniewallner** [comment]
Support for PEP 420 has been added back in [0.19.1](https://github.com/fpgmaas/deptry/releases/tag/0.19.1) behind `--experimental-namespace-package` feature flag (and the corresponding `experimental_namespace_package` option under `[tool.deptry]` in `pyproject.toml`).

Performance should be slightly better as we now do an early return if we find at least one Python file in the directory. There are some possible future improvements listed in https://github.com/fpgmaas/deptry/pull/808 to improve performance even more.

**Comment by charlesnicholson** [comment]
Hmm, I'm still seeing the error in my little repro testbed on v0.19.1:
<img width="1169" alt="Screenshot 2024-08-11 at 8 30 12 PM" src="https://github.com/user-attachments/assets/a7761e0c-4bba-4674-8036-7559d246df19">

The failed action is here:
https://github.com/charlesnicholson/deptry-pep420-repro/actions/runs/10343972148/job/28628860626

and the repro, updated to use `--experimental-namespace-package` is here:
https://github.com/charlesnicholson/deptry-pep420-repro

(Is my example incorrect?)

**Comment by mkniewallner** [comment]
> Hmm, I'm still seeing the error in my little repro testbed on v0.19.1: <img alt="Screenshot 2024-08-11 at 8 30 12 PM" width="1169" src="https://private-user-images.githubusercontent.com/3010295/356914501-a7761e0c-4bba-4674-8036-7559d246df19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1MDQxNTQsIm5iZiI6MTcyMzUwMzg1NCwicGF0aCI6Ii8zMDEwMjk1LzM1NjkxNDUwMS1hNzc2MWUwYy00YmJhLTQ2NzQtODAzNi03NTU5ZDI0NmRmMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTJUMjMwNDE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWVlODgxMzNmODk3MTliNzBiM2E3YjA2ODliYTMyYWRmODE2MGExYTY4NmQyY2QyM2Q0YTViMmMxN2M4OGI5MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2ZY5bAS5r5Mv_X5vr6jL9tAqIQVNaicvu0w8BRcDYQY">
> 
> The failed action is here: https://github.com/charlesnicholson/deptry-pep420-repro/actions/runs/10343972148/job/28628860626
> 
> and the repro, updated to use `--experimental-namespace-package` is here: https://github.com/charlesnicholson/deptry-pep420-repro
> 
> (Is my example incorrect?)

Yes sorry, nothing has changed since https://github.com/fpgmaas/deptry/issues/740#issuecomment-2211033823 for now, we've just reverted the feature because of a performance regression, then put it back behind a feature flag, but as is, meaning it doesn't fully fix your use case. I did not yet take the time to fully dig into it, but still planned at some point.
